libdmmp/libdmmp.c

```diff
@@ -0,0 +1,437 @@
+	bool _has_rlimit;
+	int log_priority;
+};
+
+_dmmp_getter_func_gen(dmmp_context_log_priority_get,
+		      struct dmmp_context, ctx, log_priority, int,
+		      DMMP_LOG_PRIORITY_DEFAULT);
+
+_dmmp_array_free_func_gen(dmmp_mpath_array_free, struct dmmp_mpath,
+			  _dmmp_mpath_free);
+
+static int _DMMP_LOG_STRERR_ALIGN_WIDTH = 80;
+/* ^ Used in _log_stderr(). If provides log message less than 80 bytes,
+ *   fill it with space, then print code file name, function name, line after
+ *   the 80th bytes.
+ */
+
+/*
+ * Do initial works:
+ *  Load config.
+ *  Apply max_fd limits.
+ */
+static int _init(struct dmmp_context *ctx);
+
+/*
+ * Clean up what _init did
+ */
+static void _close(struct dmmp_context *ctx);
+
+/*
+ * Malloc and initialize.
+ */
+
+static int _get_pathvec(struct dmmp_context *ctx, vector *pathvec);
+
+static void _log_stderr(struct dmmp_context *ctx,
+			enum dmmp_log_priority priority,
+			const char *file, int line, const char *func_name,
+			const char *format, va_list args);
+static void _update_path_status(struct multipath *mpp);
+
+static void _log_stderr(struct dmmp_context *ctx,
+			enum dmmp_log_priority priority,
+			const char *file, int line, const char *func_name,
+			const char *format, va_list args)
+{
+	int printed_bytes = 0;
+
+	printed_bytes += fprintf(stderr, "libdmmp %s: ",
+				 dmmp_log_priority_str(priority));
+	printed_bytes += vfprintf(stderr, format, args);
+
+	if (printed_bytes < _DMMP_LOG_STRERR_ALIGN_WIDTH) {
+		fprintf(stderr, "%*s # %s:%s():%d\n",
+			_DMMP_LOG_STRERR_ALIGN_WIDTH - printed_bytes, "", file,
+			func_name, line);
+	} else {
+		fprintf(stderr, " # %s:%s():%d\n", file, func_name, line);
+	}
+}
+
+static int _init(struct dmmp_context *ctx)
+{
+	struct rlimit fd_limit;
+	int rc = DMMP_OK;
+	struct udev *udev = NULL;
+
+	udev = udev_new();
+	if (udev == NULL) {
+		rc = DMMP_ERR_NO_MEMORY;
+		_error(ctx, dmmp_strerror(rc));
+		return rc;
+	}
+
+	if (load_config(DEFAULT_CONFIGFILE, udev) != 0) {
+		rc = DMMP_ERR_LOAD_CONFIG_FAIL;
+		/* TODO(Gris Ge): Update load_config() to provide better error
+		 *		  message.
+		 */
+		_error(ctx, "%s: %s", dmmp_strerror(rc), DEFAULT_CONFIGFILE);
+		goto out;
+	}
+	_debug(ctx, "Config %s loaded", DEFAULT_CONFIGFILE);
+
+	if (conf->max_fds) {
+		/* Save current limit and restore it at _close() */
+		if (getrlimit(RLIMIT_NOFILE, &ctx->_rlimit) != 0) {
+			rc = DMMP_ERR_BUG;
+			_error(ctx, "Failed to get current "
+			       "RLIMIT_NOFILE limit: %d:%s", errno,
+			       strerror(errno));
+			goto out;
+		}
+
+		fd_limit.rlim_cur = conf->max_fds;
+		fd_limit.rlim_max = conf->max_fds;
+		if (setrlimit(RLIMIT_NOFILE, &fd_limit) != 0)
+			_warn(ctx, "can't set open fds limit to %d : %s",
+				conf->max_fds, strerror(errno));
+	}
+
+out:
+	if (udev != NULL)
+		udev_unref(udev);
+
+	if (rc != DMMP_OK)
+		_close(ctx);
+	return rc;
+}
+
+static void _close(struct dmmp_context *ctx)
+{
+	dm_lib_release();
+	dm_lib_exit();
+
+	cleanup_prio();
```

Comment:
Is this necessary?  The code doesn't initialize the the checkers/prioritizers, right?